dace/frontend/python/replacements.py

```diff
@@ -0,0 +1,789 @@
+                 op2, ','.join(['__i%d' % i for i in range(len(arr1.shape))]))
+         },
+        '__out = __in1 %s __in2' % opcode, {
+            '__out':
+            Memlet.simple(
+                name, ','.join(['__i%d' % i for i in range(len(arr1.shape))]))
+        },
+        external_edges=True)
+    return name
+
+
+def _scalarbinop(sdfg: SDFG,
+                 state: SDFGState,
+                 scalop: str,
+                 arrop: str,
+                 opcode: str,
+                 opname: str,
+                 restype: dace.typeclass,
+                 reverse: bool = False):
+    """ Implements a general Scalar-Array binary operator. """
+    scalar = sdfg.arrays[scalop]
+    arr = sdfg.arrays[arrop]
+
+    name, _ = sdfg.add_temp_transient(arr.shape, restype, arr.storage)
+    state.add_mapped_tasklet(
+        "_SA%s_" % opname,
+        {'__i%d' % i: '0:%s' % s
+         for i, s in enumerate(arr.shape)}, {
+             '__in1':
+             Memlet.simple(scalop, '0'),
+             '__in2':
+             Memlet.simple(
+                 arrop, ','.join(['__i%d' % i
+                                  for i in range(len(arr.shape))])),
+         },
+        '__out = %s %s %s' % ('__in2' if reverse else '__in1', opcode,
+                              '__in1' if reverse else '__in2'),
+        {
+            '__out':
+            Memlet.simple(
+                name, ','.join(['__i%d' % i for i in range(len(arr.shape))]))
+        },
+        external_edges=True)
+    return name
+
+
+# Defined as a function in order to include the op and the opcode in the closure
+def _makeassignop(op, opcode):
+    @oprepo.replaces_operator('Array', op)
+    def _op(visitor: 'ProgramVisitor',
+            sdfg: SDFG,
+            state: SDFGState,
+            op1: str,
+            op2=None):
+        return _assignop(sdfg, state, op1, opcode, op)
+
+
+def _makeunop(op, opcode):
+    @oprepo.replaces_operator('Array', op)
+    def _op(visitor: 'ProgramVisitor',
+            sdfg: SDFG,
+            state: SDFGState,
+            op1: str,
+            op2=None):
+        return _unop(sdfg, state, op1, opcode, op)
+
+
+@oprepo.replaces_operator('int', 'USub', None)
+@oprepo.replaces_operator('float', 'USub', None)
+def _neg(visitor: 'ProgramVisitor',
+         sdfg: SDFG,
+         state: SDFGState,
+         op1: Union[int, float],
+         op2=None):
+    return -op1
+
+
+@oprepo.replaces_operator('symbol', 'Add', 'int')
+@oprepo.replaces_operator('symbol', 'Add', 'float')
+def _addsym(visitor: 'ProgramVisitor', sdfg: SDFG, state: SDFGState,
+            op1: symbolic.symbol, op2: Union[int, float]):
+    return op1 + op2
+
+
+def _is_scalar(sdfg: SDFG, arrname: str):
+    """ Checks whether array is pseudo-scalar (shape=(1,)). """
+    shape = sdfg.arrays[arrname].shape
+    if len(shape) == 1 and shape[0] == 1:
+        return True
+    return False
+
+
+def _inverse_dict_lookup(dict: Dict[str, Any], value: Any):
```

Comment:
so this function appears twice now. Yet another reason for common.py with that function too